### PR TITLE
Input filter and behavior suggestions [finishes #177453563]

### PR DESCRIPTION
**Pull request by lucaskuzma**

### CODE REVIEW

File: FantasmoSDK/Classes/Filters/FMBlurFilter.swift [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+//  Created by lucas kuzma on 3/30/21.
+//
+
+import ARKit
+import Metal
+import MetalPerformanceShaders
+import MetalKit
+import VideoToolbox
+
+class FMBlurFilter: FMFrameFilter {
```
Comment (yklishevich) [inline FantasmoSDK/Classes/Filters/FMBlurFilter.swift:14]: Filter must be reset when new sequence of frames starts ( on ```startUpdatingLocation/stopUpdatingLocation```)

**Review by yklishevich — CHANGES_REQUESTED** [review]

File: FantasmoSDK/Classes/Filters/FMInputQualityFilter.swift [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    func accepts(_ frame: ARFrame) -> Bool {
+
+        if !shouldForceAccept() {
+            // run frame through filters
+            for filter in filters {
+                if case let .rejected(rejection) = filter.accepts(frame) {
+                    addRejection(rejection)
+                    notifyIfNeeded(rejection)
+                    return false
+                }
```
Comment (yklishevich) [inline FantasmoSDK/Classes/Filters/FMInputQualityFilter.swift:71]: Our plain (non compound) filters (filter rules) are all "sequence" filters, that is each filter supposes that it will receive all frames from sequence. This is important for proper work of "sequence filter". So, even if some plain filter fails we should also pass frame to all other plain filters to ensure that no frame was dropped for some plain filter.

File: FantasmoSDK/Classes/Filters/FMBlurFilter.swift [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        metalCommandBuffer.commit()
+        metalCommandBuffer.waitUntilCompleted()
+        
+        // grab results
+        var result = [Int8](repeatElement(0, count: 2))
+        let region = MTLRegionMake2D(0, 0, 2, 1)
+        varianceTexture.getBytes(&result, bytesPerRow: 1 * 2 * 4, from: region, mipmapLevel: 0)
+        
+        return Int(result.last ?? 0)
+    }
```
[the remark this answers is not in the document]
Comment (lucaskuzma) [reply FantasmoSDK/Classes/Filters/FMBlurFilter.swift:106]: I am seeing variance values as high as 11 when viewing feature-rich images. Are you testing with outdoor scenery or a relatively featureless office interior?

[the remark this answers is not in the document]
Comment (lucaskuzma) [reply FantasmoSDK/Classes/Filters/FMBlurFilter.swift:106]: Can you summarize what is different in your implementation?

File: FantasmoSDK/Classes/Filters/FMInputQualityFilter.swift (hunk 2 above)
Comment (lucaskuzma) [reply FantasmoSDK/Classes/Filters/FMInputQualityFilter.swift:76]: In the case of the two stateful filters, we do not care if we capture every frame. Afaict they deliver desired results even when considering only the frames that make it through the previous filters. Your point is a good one, and we should consider it if we add a filter that really does need every frame. The current design is intended to minimize computational cost, particularly that incurred by the blur filter.
